starter/privileged_service_client.h

```diff
@@ -0,0 +1,53 @@
+#pragma once
+
+#if !defined(__linux__)
+#error "Linux platform file is part of non-Linux build."
+#endif
+
+#include "source/common/common/assert.h"
+#include "source/common/singleton/threadsafe_singleton.h"
+
+#include "starter/privileged_service_protocol.h"
+
+namespace Envoy {
+namespace Cilium {
+
+class Bpf;
+class SocketMarkOption;
+
+namespace PrivilegedService {
+
+// ProtocolClient implements the client logic for communicating with the privileged service.
+class ProtocolClient : public Protocol {
+public:
+  ProtocolClient();
+
+  // allow access to the classes that need it
+  friend class Envoy::Cilium::Bpf;
+  friend class Envoy::Cilium::SocketMarkOption;
+
+protected:
+  // Read-only bpf syscalls
+  Envoy::Api::SysCallIntResult bpf_open(const char *path);
```

Comment:
document the 3 "public" facing functions?

Reply:
I'm thinking this privileged API will morph to something more CIlium specific in near future. Giving access to set any/all socket options and open any bpf/all bpf maps is not the most secure thing to do. Both the bpf lookup and socket option setting happen via the bpf_metadata filter, so there would be an opportunity to lift the abstraction level of this privileged API from the current syscall level to something like "Cilium connection metadata setup" that would:
- Get the metadata for the new connection from conntrack and/or ipcache maps, and, if successful
- Set the socket options Cilium needs for new proxied connections
A separate call would be needed to set up the listening socket (socket options only)

This is a bit bigger change, shifting all the bpf code to the starter and refactoring the bpf_metadata filter code, so I did not want to start with this "bigger picture".

So in summary, I'd like to skip documenting this version of the privileged API, and would like to not characterise this as "public facing", even though it kind of is.

Reply:
makes sense. thanks for the outlook and explanation!